core/models/sessions.go

```diff
@@ -769,7 +769,73 @@ func WriteSessionOutputsToStorage(ctx context.Context, rt *runtime.Runtime, orgI
 	return nil
 }
 
-// ExitSessions exits sessions and their runs. It batches the given session ids and exits each batch in a transaction.
+// InterruptSessions interrupts waiting sessions and their runs
+func InterruptSessions(ctx context.Context, db *sqlx.DB, uuids []flows.SessionUUID) error {
+	// split into batches and exit each batch in a transaction
+	for batch := range slices.Chunk(uuids, 100) {
+		tx, err := db.BeginTxx(ctx, nil)
+		if err != nil {
+			return fmt.Errorf("error starting transaction to interrupt sessions: %w", err)
+		}
+
+		if err := interruptSessionBatch(ctx, tx, batch); err != nil {
+			return fmt.Errorf("error interrupting batch of sessions: %w", err)
+		}
+
+		if err := tx.Commit(); err != nil {
+			return fmt.Errorf("error committing session interrupts: %w", err)
+		}
+	}
+
+	return nil
+}
+
+const sqlInterruptSessions = `
+   UPDATE flows_flowsession
+      SET status = 'I', ended_on = NOW(), current_flow_id = NULL
+    WHERE uuid = ANY ($1) AND status = 'W'
+RETURNING contact_id`
+
+// TODO instead of having an index on session_uuid.. rework this to fetch the sessions and extract a list of run uuids?
+const sqlInterruptSessionRuns = `
+UPDATE flows_flowrun
+   SET exited_on = NOW(), status = 'I', modified_on = NOW()
+ WHERE session_uuid = ANY($1) AND status IN ('A', 'W')`
+
+const sqlInterruptSessionContacts = `
+ UPDATE contacts_contact 
+    SET current_session_uuid = NULL, current_flow_id = NULL, modified_on = NOW() 
+  WHERE id = ANY($1) AND current_session_uuid = ANY($2)`
```

Comment:
Usually when we have a set of things to update where A = X AND B = Y we do this complicated temp table thing but these are two discreet sets.. so this seems safe... and certainly simpler.